src/examples/algorithms/Anomaly.cpp

```diff
@@ -0,0 +1,199 @@
+#include <iostream>     
+#include <fstream>      
+#include <vector>
+#include <string>
+#include <algorithm>    
+#include <iterator>     
+#include <boost/tokenizer.hpp>
+
+#include "nupic/encoders/ScalarEncoder.hpp"
+#include "nupic/algorithms/SpatialPooler.hpp"
+#include "nupic/algorithms/Cells4.hpp"
+#include "nupic/algorithms/Anomaly.hpp"
+#include "nupic/os/Timer.hpp"
+
+
+// Prints a vector as long as it's type has operator << defined.
+template<typename T>
+void print_vector(const std::vector<T>& v, char sep=0)
+{
+    for (auto& elem: v) {
+        std::cout << elem;
+        if (sep != 0) {
+            std::cout << sep;
+        }
+    }
+    std::cout << std::endl;
+}
+
+
+// Converts (casts) all elements of a vector from T1 to T2 (returns a new vector)
+template<typename T1, typename T2>
+std::vector<T2> vectorTypeCast(const std::vector<T1>& orig)
+{
+    std::vector<T2> dest;
+    std::transform(orig.cbegin(), orig.cend(), std::back_inserter(dest),
+                   [](const T1& elem) { return static_cast<T2>(elem); });
+    return dest;
+}
+
+
+// Converts an array of 1s and 0s to a sparse vector
+std::vector<UInt> binaryToSparse(const std::vector<UInt>& v)
+{
+    std::vector<UInt> sparse;
+    for (UInt index = 0; index < v.size(); index++) {
+        if (v[index] > 0) { // generic form, so values like 0.3 works as well
+            sparse.push_back(index);
+        }
+    }
+    return sparse;
+}
+
+
+// Converts the output of the temporal pooler (columns of 4 cells) to a single-cell-per-column
+// representation
+std::vector<UInt> normalizeTPOutput(const std::vector<UInt>& v, UInt cellsPerColumn)
+{
+    std::vector<UInt> normalized;
+    for (unsigned int i = 0; i < v.size(); ) {
+        unsigned int val = 0;
+        unsigned int j;
+        for (j = i; j < v.size() && j < i + cellsPerColumn; j++) {
+            val |= v[j];
+        }
+        normalized.push_back(val);
+        i = j;
+    }
+    return normalized;
+}
+
+// the chain of: input -> encoder -> spatial pooler -> temporal memory -> anomaly -> output
+int main()
+{
+    using namespace nupic;
+    using namespace nupic::algorithms::spatial_pooler;
+    using namespace nupic::algorithms::Cells4;
+    using namespace nupic::algorithms::anomaly;
+
+    std::string data("data.csv");
```

Comment:
FIXME: use `getenv` and `$NUPIC_CORE` to get absolute path; where should we place the example data (examples/data/ ?) 
Or, parse command line argument for an absolute path to the data.